Add support for sun.stderr.encoding and sun.stdout.encoding

pshipton opened the pull request:
Also support setting extended Charset's in the jdk.charsets module.

See https://github.com/eclipse/openj9/pull/11387

Signed-off-by: Peter Shipton <Peter_Shipton@ca.ibm.com>

pshipton commented:
@JasonFengJ9 @Norimakisushi @narkedi pls review

I'll see about adding a test, otherwise this is ready for review.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -23,6 +23,9 @@
 
 import java.io.*;
+/*[IF Sidecar18-SE-OpenJ9 | JAVA_SPEC_VERSION >= 11]*/
+import java.nio.charset.*;
```
I'm not a fan of wildcard imports; please import explicitly:
```
import java.nio.charset.Charset;
import java.nio.charset.IllegalCharsetNameException;
import java.nio.charset.UnsupportedCharsetException;
```

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -101,6 +104,11 @@
 	private static String platformEncoding;
 	private static String fileEncoding;
 	private static String osEncoding;
+	
```
Please no new trailing whitespace.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -101,6 +104,11 @@
 	private static String osEncoding;
+	
+	/*[IF JAVA_SPEC_VERSION >= 11]*/
+	static boolean setErrEncoding;
+	static boolean setOutEncoding;
```
These names could be better, rather than leave the reader guessing whether the meaning is 'did set' or 'should set'.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -133,6 +141,37 @@ private static void initEncodings() {
+				} else {
+					return newCharset.name();
+				}
+			} catch (IllegalCharsetNameException | UnsupportedCharsetException e) {
+			}
```
An explicit comment explaining why the exceptions are being ignore would be welcome.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -176,9 +215,33 @@ static void afterClinitInitialization() {
+				setOut(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.out)), true));
+			}
+		} catch (UnsupportedEncodingException e) {
+			// not possible as the Charset name was validated first
```
I think it's better to assert that by throwing `InternalError` or similar.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/Thread.java:
```diff
@@ -215,6 +218,43 @@ void completeInitialization() {
+		try {
+			System.err.flush();
+			/*[IF PLATFORM-mz31|PLATFORM-mz64]*/
+			System.setOut(com.ibm.jvm.io.ConsolePrintStream.localize(new BufferedOutputStream(new FileOutputStream(FileDescriptor.out)), true, stderrCharset));
+			/*[ELSE]*/
+			System.setOut(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.out)), true, stderrCharset));
```
These should be calls to `setErr()` and should use `FileDescriptor.err`.

pshipton commented:
I think I've fixed all the review comments.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -132,7 +142,40 @@ private static void initEncodings() {
 			osEncoding = definedOSEncoding;
 	}	
-		
+
```
No trailing whitespace, please.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -132,7 +142,40 @@ private static void initEncodings() {
+				}
+			} catch (IllegalCharsetNameException | UnsupportedCharsetException e) {
+				// ignore unsupported or invalid encodings
+			}		
```
No trailing whitespace, please.

JasonFengJ9 commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -132,7 +142,40 @@ private static void initEncodings() {
+	/*
+	 * Return the Charset name of the primary or fallback encoding, if different from the default.
+	 */
+	static String getCharsetName(String primary, String fallback, Charset defaultCharset) {
```
Is there any planned usage for `fallback` other than `props.getProperty("file.encoding")` and `null`?
The reason I am asking is that `Charset.defaultCharset()` gets its `Charset` via same system property `file.encoding` as well hence I suspect the fallback is same as defaultCharset which is not necessary.

keithc-ca replied:
I think Jason is correct. In System, the fallback name is null. The calls from Thread should match the default charset.
I think the fallback parameter should simply be removed.

keithc-ca replied:
If this method returned the Charset rather than it's name, it would simplify the code where it's used.

keithc-ca replied:
We can do that for Java 11, but not Java 8 (the `PrintStream` constructor we would want to use was added in Java 10).

pshipton replied:
I'll add more comments to better explain why we need to use file.encoding in Thread.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -176,9 +219,39 @@ static void afterClinitInitialization() {
 		StringCoding.encode(new char[1], 0, 1);
 		/*[ENDIF]*/
+
 		/*[IF (Sidecar18-SE-OpenJ9|Sidecar19-SE)&!(PLATFORM-mz31|PLATFORM-mz64)]*/
```
The code below would be easier to understand if this condition were repeated in the matching `[ENDIF]` comment.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/Thread.java:
```diff
@@ -215,6 +222,43 @@ void completeInitialization() {
+	}
+
+	// If the sun.stdout.encoding was already set in System, don't set it again
+	String stdoutEncoding = System.hasSetOutEncoding ? null : System.internalGetProperties().getProperty("sun.stdout.encoding"); //$NON-NLS-1$
```
This should use `props` instead of calling `System.internalGetProperties()` again.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/Thread.java:
```diff
@@ -215,6 +222,43 @@ void completeInitialization() {
+			/*[ELSE]*/
+			System.setOut(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.out)), true, stdoutCharset));
+			/*[ENDIF] PLATFORM-mz31|PLATFORM-mz64 */
+		} catch (UnsupportedEncodingException e) {
```
Please either add a comment here (and for line 243) explaining why it's safe to ignore the exception, or, better, throw an InternalError.

pshipton commented:
Addressed the latest review comments.

pshipton commented:
Fixed some comment grammar.

pshipton commented:
Updated the commit comment.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -131,8 +141,62 @@ private static void initEncodings() {
+	/*
+	 * Return the Charset of the primary or fallback encoding, if different from the default.
+	 */
+	static Charset getCharsetName(String primary, String fallback, Charset defaultCharset) {
```
`getCharset` would be a more accurate name for this method.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -131,8 +141,62 @@ private static void initEncodings() {
+	/*
+	 * Return the Charset name of the primary encoding, if different from the default.
+	 */
+	static String getCharsetName(String primary, Charset defaultCharset) {
```
This can be private.

pshipton commented:
Added tests.

pshipton commented:
Addressed latest review comments.

pshipton commented:
jenkins compile zlinux jdk8,jdk11,jdk16

pshipton commented:
Tested the new test via grinders, which passed.
jdk8: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1642
jdk11: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1643
jdk16: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1644

pshipton commented:
Updated the z/OS support, and some cosmetic changes. The tests still need updating for z/OS.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -102,6 +105,14 @@
 	private static String osEncoding;
 
+	/*[IF JAVA_SPEC_VERSION >= 11]*/
+	static boolean hasSetErrEncoding;
+	static boolean hasSetOutEncoding;
+	/*[IF PLATFORM-mz31|PLATFORM-mz64]*/
+	static Charset consoleDefaultCharset;
```
These fields should be `private` or the accessor methods could be removed.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -176,18 +255,83 @@ static void afterClinitInitialization() {
+		Properties props = internalGetProperties();
+		Charset consoleCharset = Charset.defaultCharset();
+		/*[IF PLATFORM-mz31|PLATFORM-mz64]*/
+		String systemEncoding = props.getProperty("ibm.system.encoding");
```
This variable could be inlined like other property accesses below.
Also, the `//$NON-NLS-1$` comment is missing.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/Thread.java:
```diff
@@ -215,6 +221,47 @@ void completeInitialization() {
+	 * in the java.base module and so are not used at startup. There are additional Charset's in the
+	 * jdk.charsets module, which is only loaded later. This means the default Charset may not be the
+	 * same as file.encoding. Now that all modules and Charset's are available, check if the desired
+	 * encodings can be used for System.err and System.out. 
```
Please no trailing whitespace.

keithc-ca commented on test/functional/cmdLineTests/cmdLineTest_J9tests/j9tests_console.xml:
```diff
@@ -0,0 +1,76 @@
+  <return type="success" value="0" />
+ </test>
+ 
+  <test id="file.encoding System.out $ENCODING2$">
```
Inconsistent indentation.
Also, why just one space per level (it makes it a little harder to read)? How about two spaces per level?

keithc-ca commented on test/functional/cmdLineTests/cmdLineTest_J9tests/playlist.xml:
```diff
@@ -140,4 +140,51 @@
+			<impl>openj9</impl>
+		</impls>
+	</test>
+	<test>
+		<testCaseName>cmdLineTest_J9test_console_zos</testCaseName>
+		<command>$(JAVA_COMMAND) $(JVM_OPTIONS) \
+	-DJARPATH=$(Q)$(TEST_RESROOT)$(D)cmdLineTest_J9tests.jar$(Q) -DJDK_VERSION=$(JDK_VERSION) \
+	-DJVMLIBPATH=$(Q)$(J9VM_PATH)$(Q) \
+	-DTESTDIR=$(Q)$(TEST_RESROOT)$(Q) -DRESJAR=$(CMDLINETESTER_RESJAR) -DEXE=$(SQ)$(JAVA_COMMAND) $(JVM_OPTIONS) -Xdump$(SQ) \
+	-jar $(CMDLINETESTER_JAR) \
+	-config $(Q)$(TEST_RESROOT)$(D)j9tests_console.xml$(Q) \
+	-xids all,$(PLATFORM) -plats all,$(PLATFORM) -xlist $(Q)$(TEST_RESROOT)$(D)j9tests_exclude.xml$(Q) \
+	-explainExcludes -nonZeroExitWhenError; \
+	$(TEST_STATUS)</command>
+		<platformRequirements>os.zos</platformRequirements>
+		<levels>
+			<level>extended</level>
+		</levels>
+		<groups>
+			<group>functional</group>
+		</groups>
+		<versions>
+			<version>11+</version>
+		</versions>
+		<impls>
+			<impl>ibm</impl>
+			<impl>openj9</impl>
+		</impls>
+	</test>
```
How does this test differ from the one above (other than applying only for version >= 11)?
Why is the first test insufficient?

pshipton replied:
It is the same test, but applied to z/OS Java 11+ only. The tests apply to builds other than IBM Java 8, the test above is only for `<impl>openj9</impl>` which doesn't cover z/OS Java 11+.

keithc-ca replied:
The test above will apply to zos as well as it stands. Do we need or want that duplication?
Do we want the one above to say
```
<platformRequirements>^os.zos</platformRequirements>
```

pshipton replied:
I don't believe the test above does apply to zOS, as it uses only `<impl>openj9</impl>` but z/OS is "ibm". @llxia is this the best way to add tests everywhere but IBM Java 8?

keithc-ca replied:
Yes, I suppose we don't expect an 'openj9' implementation for z/OS, in which case it would not apply.

pshipton replied:
To confirm, I tried running the non-zos test on z/OS, but no tests ran as expected.

llxia replied:
> is this the best way to add tests everywhere but IBM Java 8? 

Yes, I think so.

keithc-ca commented on test/functional/cmdLineTests/cmdLineTest_J9tests/src/GetConsoleCharset.java:
```diff
@@ -0,0 +1,61 @@
+ */
+
+public class GetConsoleCharset {
+    public static void main(String[] args) throws Exception {
+    	OutputStream stream;
+    	if (args.length < 2) {
+    		System.out.println("Provide two arguments, the first is either 'out' or 'err', and the second is the expected Charset name.");
+    		System.exit(1);
+    	}
+        if ("out".equals(args[0])) {
+        	stream = System.out;
+        } else {
+        	stream = System.err;
+        }
+        Field charOutField = PrintStream.class.getDeclaredField("charOut");
+        charOutField.setAccessible(true);
+        String encoding = ((OutputStreamWriter)charOutField.get(stream)).getEncoding();
+        System.out.println("System." + args[0] + " encoding: " + encoding);
+        if (args.length > 2) {
+        	FileOutputStream out = new FileOutputStream("GetConsoleCharset-result.txt");
+        	out.write(encoding.getBytes("ISO-8859-1"));
+        	out.close();
+        }
+        if (encoding.equals(args[1])) {
+        	System.out.println("SUCCESS");
+        	System.exit(0);
+        }
+    	System.out.println("FAILED");
+        System.exit(1);
+    }
```
Please indent consistently (just tabs, not a combination of both).

Norimakisushi commented:
Although, it was tested via the grinder, some more test cases were run on Linux-x86_64.  (Sorry, not on z/OS).
These tests were used in #11387. They compare the behavior of the openjdk8 and 11 with hotspot. 
It was confirmed that now sun.stdout.encoding and sun.stderr.encoding behave compatible with them.

Please refer the attached spread sheet for good test results.

There is one point which may tern to be a concern in future;
Openjdk 11 and later's defaultCharset has an issue. Not like openjdk8, some charsets could not be set as default.
It's cause is written in #11387 as the cause of sun.stdout.encoding become fall-back charsets.
It is being considered to fix if the defaultCharset will be set again at JVM post-initialization.

Currently, when fall-back charset is set by defaultCharset(), next statements returns fall-back charset. 
jcl/src/java.base/share/classes/java/lang/Thread.java 
237	Charset consoleCharset = Charset.defaultCharset();
If defulatCharset() could be  fixed as considered, this statement will return specified Charset, even fall-back is assigned during JVM initialization.
It may results a skip of necessary 2nd setOut, setErr.
Is it possible to around this situation? For example, call defaultCharset) within getCharset() first time during initialization and keep it within getCharset().

[defaultCharset-test-results-0405.xlsx](https://github.com/eclipse/openj9/files/6259339/defaultCharset-test-results-0405.xlsx)

pshipton commented:
Addressed all the review comments, and fixed a logic problem in Thread which would have reset the console encoding set via `sun.std<err/out>.encoding` that was already set in System, if the file.encoding Charset wasn't available at startup.

Build: https://ci.eclipse.org/openj9/job/Pipeline-Build-Test-Personal/70/
Tests jdk8: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1647/
jdk11: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1648/

JasonFengJ9 commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -176,18 +251,82 @@ static void afterClinitInitialization() {
+		/*[IF JAVA_SPEC_VERSION >= 11]*/
+		if (stderrCharset != null) {
+			hasSetErrEncoding = true;
+			setErr(com.ibm.jvm.io.ConsolePrintStream.localize(new BufferedOutputStream(new FileOutputStream(FileDescriptor.err)), true, stderrCharset));
```
Since `Add ConsolePrintStream.localize with encoding` within JDK11 zOS confirmed that `com.ibm.jvm.io.ConsolePrintStream` is only required when `line.separator` is not `\n`.
I think a check can be added to limit the usage of `com.ibm.jvm.io.ConsolePrintStream` at the case that `line.separator` is not `\n` due to platform specific or command line settings.

pshipton replied:
> Since Add ConsolePrintStream.localize with encoding within JDK11 zOS confirmed that com.ibm.jvm.io.ConsolePrintStream is only required when line.separator is not \n.

I don't think is true. ConsolePrintStream does a conversion when the file.encoding is different from the ibm.system.encoding.

JasonFengJ9 replied:
> ConsolePrintStream does a conversion when the file.encoding is different from the ibm.system.encoding.

As per `ConsolePrintStream` description:
```
 *     Wrapper class to perform NL to line.separator conversion on platforms
 *     where the line.separator != "\n". Intended for use on System.out and
 *     System.err .
```
If that's its sole purpose, there is no need to use `ConsolePrintStream` when current `line.separator == \n`.
There are conversions other than `\n` to `line.separator`, my question is, are they really needed? It might worth the effort to understand it and avoid the overhead of using `ConsolePrintStream` if possible.

pshipton replied:
> There are conversions other than `\n` to `line.separator`, my question is, are they really needed? 

Somebody thought so, but it's not clear what the use cases are. We'll have to try and get that information from the IBM JCL team.

> It might worth the effort to understand it and avoid the overhead of using ConsolePrintStream if possible.

I expect so, but it seems unrelated to adding support for sun.stderr.encoding, sun.stdout.encoding and can be done separately. Considering IBM Java 8 uses it for all platforms, it seems very risky to consider removing it until all the use cases are understood. Removing it for Java 11 z/OS may be possible. Considering the ConsolePrintStream impl isn't part of OpenJ9, I don't think this is the right place to discuss it. Already you are copying comments from restricted materials to open source comments.

JasonFengJ9 replied:
> Removing it for Java 11 z/OS may be possible.

Yeah, that's what I was aiming for.
The comments referred I thought are from JCL team like other internal docs (RTC contents, etc.), in addition, this class is in IBM Java 8 all platforms.

keithc-ca commented on test/functional/cmdLineTests/cmdLineTest_J9tests/j9tests_console.xml:
```diff
@@ -0,0 +1,78 @@
+  </test>
+
+  <test id="sun.stdout.encoding $ENCODING1$">
+    <command>$EXE$ $OPENS$ $CP$ -D$DEFAULT_ENCODING_PROP$=$ENCODING2$ -Dsun.stdout.encoding=$ENCODING1$ $TARGET$ out $ENCODING1$</command>
```
This should use `ENCODING1` to match the test id.

pshipton replied:
It does use `ENCODING1` for the `sun.stdout.encoding` and expects the result to be `ENCODING1`. Howver it also sets `DEFAULT_ENCODING_PROP` (i.e. `file.encoding`) to `ENCODING2` expecting this to be ignored (which it is).

keithc-ca replied:
Got it; thanks.

keithc-ca commented on test/functional/cmdLineTests/cmdLineTest_J9tests/j9tests_console.xml:
```diff
@@ -0,0 +1,78 @@
+    <return type="success" value="0" />
+  </test>
+
+  <test id="sun.stderr.encoding $ENCODING1$">
+    <command>$EXE$ $OPENS$ $CP$ -D$DEFAULT_ENCODING_PROP$=$ENCODING2$ -Dsun.stderr.encoding=$ENCODING1$ $TARGET$ err $ENCODING1$</command>
```
The id should use `ENCODING2` to match the test command.

pshipton replied:
See the response to the previous comment.

JasonFengJ9 commented on jcl/src/java.base/share/classes/java/lang/System.java:
```diff
@@ -176,18 +251,82 @@ static void afterClinitInitialization() {
+		}
+		if (stdoutCharset != null) {
+			hasSetOutEncoding = true;
+			setOut(com.ibm.jvm.io.ConsolePrintStream.localize(new BufferedOutputStream(new FileOutputStream(FileDescriptor.out)), true, stdoutCharset));
```
Now there are quite a few references of `com.ibm.jvm.io.ConsolePrintStream`, it can be imported.
Another suggestion is to make a helper method for `BufferedOutputStream` instantiation to reduce code duplication.

pshipton replied:
I'll implement these in a separate PR.

keithc-ca commented:
Jenkins test sanity,extended osx,plinux,win jdk8,jdk11

keithc-ca commented:
Jenkins test sanity,extended win jdk8
